src/GPU3D.cpp

```diff
@@ -546,7 +550,171 @@ void GPU3D::SetEnabled(bool geometry, bool rendering) noexcept
     if (!rendering) ResetRenderingState();
 }
 
+void GPU3D::NewWriteFD(u8 cmd, u32* param)
+{
+    // note: 0dotdisp is assigned an id of 0x72, to avoid having to do any special handling for it, since vec and box tests aren't saved.
+    if (!NDS.GPU.FDInProg || (FD->NumParams >= sizeof(FD->Params) / sizeof(FD->Params[0])) || (FD->NumCmds >= sizeof(FD->Cmd) / sizeof(FD->Cmd[0]))) return;
+
+    FD->Cmd[FD->NumCmds++] = cmd;
+
+    for (int i = 0; i < CmdNumParams[cmd]; i++)
+        FD->Params[FD->NumParams++] = param[i];
+}
+
+void GPU3D::StartFrameDump()
+{
+    NDS.GPU.QueueFrameDump = false;
+    FD = std::make_unique<FrameDump>();
```

Comment:
If `make_unique` fails due to running out of memory then it won't return `nullptr`. It'll throw an exception. If we're out of memory there's not much to do besides crash anyway.

Reply:
Is there a way to handle this without crashing or do we just accept our fate?

Reply:
Probably just have to accept your fate. On paper, you could handle it with a `try`/`catch` statement like any other exception. But in practice, I'm not sure. Here's why:

Some operating systems (such as Linux) over-commit memory, i.e. they may give you an address space but not allocate memory until you try to use it. A consequence of this is that it's possible to get a technically-valid pointer that causes a crash when you try to use it. In that case, the OS lied to you about the memory it could offer.

`bad_alloc` can supposedly be handled like an ordinary exception, but I don't know how well it plays with over-committed memory. There's probably a trick you could do with [the implicit second template argument](https://en.cppreference.com/w/cpp/memory/unique_ptr), but I don't think it's worth it in this case.

Reply:
we don't really handle failed allocations anywhere except for the rom in memory. I don't think it's a state worth recovering from for most applications (besides saving work if  possible).